File: backend/app/core/sentry.py
"""
Sentry integration for CareBow backend error monitoring.
"""
import logging
from typing import Optional
import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.redis import RedisIntegration
from sentry_sdk.integrations.logging import LoggingIntegration
from sentry_sdk.integrations.httpx import HttpxIntegration

from app.core.config import settings

logger = logging.getLogger(__name__)


def init_sentry() -> None:
    """Initialize Sentry for error monitoring and performance tracking."""
    
    if not settings.SENTRY_DSN:
        logger.info("Sentry DSN not configured, skipping Sentry initialization")
        return
    
    # Configure logging integration
    sentry_logging = LoggingIntegration(
        level=logging.INFO,        # Capture info and above as breadcrumbs
        event_level=logging.ERROR  # Send errors as events
    )
    
    # Initialize Sentry with comprehensive integrations
    sentry_sdk.init(
        dsn=settings.SENTRY_DSN,
        environment=settings.ENVIRONMENT,
        release=settings.VERSION,
        
        # Integrations for different services
        integrations=[
            FastApiIntegration(auto_enabling_integrations=False),
            SqlalchemyIntegration(),
            RedisIntegration(),
            HttpxIntegration(),
            sentry_logging,
        ],
        
        # Performance monitoring
        traces_sample_rate=1.0 if settings.ENVIRONMENT == "development" else 0.1,
        
        # Error sampling
        sample_rate=1.0,
        
        # Additional configuration
        attach_stacktrace=True,
        send_default_pii=False,  # Don't send personally identifiable information
        
        # Custom tags
        before_send=before_send_filter,
        
        # Set user context
        before_send_transaction=before_send_transaction_filter,
    )
    
    # Set global tags
    sentry_sdk.set_tag("service", "carebow-backend")
    sentry_sdk.set_tag("version", settings.VERSION)
    
    logger.info(f"Sentry initialized for environment: {settings.ENVIRONMENT}")


def before_send_filter(event, hint):
    """Filter and modify events before sending to Sentry."""
    
    # Don't send certain types of errors in development
    if settings.ENVIRONMENT == "development":
        # Skip database connection errors in development
        if "database" in str(event.get("exception", {}).get("values", [{}])[0].get("type", "")).lower():
            return None
    
    # Add custom context
    event.setdefault("tags", {})
    event["tags"]["component"] = "backend"
    
    # Add user context if available (without PII)
    if "user" in event.get("contexts", {}):
        user_context = event["contexts"]["user"]
        # Remove sensitive information
        user_context.pop("email", None)
        user_context.pop("ip_address", None)
    
    return event


def before_send_transaction_filter(event, hint):
    """Filter performance transactions before sending to Sentry."""
    
    # Don't track health check endpoints in production
    if settings.ENVIRONMENT == "production":
        transaction_name = event.get("transaction", "")
        if transaction_name in ["/health", "/", "/api/v1/health"]:
            return None
    
    return event


def capture_exception_with_context(
    exception: Exception,
    user_id: Optional[int] = None,
    extra_context: Optional[dict] = None
) -> None:
    """Capture exception with additional context."""
    
    with sentry_sdk.push_scope() as scope:
        # Add user context
        if user_id:
            scope.set_user({"id": user_id})
        
        # Add extra context
        if extra_context:
            for key, value in extra_context.items():
                scope.set_extra(key, value)
        
        # Capture the exception
        sentry_sdk.capture_exception(exception)


def capture_message_with_context(
    message: str,
    level: str = "info",
    user_id: Optional[int] = None,
    extra_context: Optional[dict] = None
) -> None:
    """Capture a message with additional context."""
    
    with sentry_sdk.push_scope() as scope:
        # Add user context
        if user_id:
            scope.set_user({"id": user_id})
        
        # Add extra context
        if extra_context:
            for key, value in extra_context.items():
                scope.set_extra(key, value)
        
        # Capture the message
        sentry_sdk.capture_message(message, level=level)


def set_user_context(user_id: int, email: Optional[str] = None) -> None:
    """Set user context for current scope."""
    user_data = {"id": user_id}
    if email and settings.ENVIRONMENT == "development":
        # Only include email in development
        user_data["email"] = email
    
    sentry_sdk.set_user(user_data)


def add_breadcrumb(message: str, category: str = "custom", level: str = "info", data: Optional[dict] = None) -> None:
    """Add a breadcrumb for debugging context."""
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        level=level,
        data=data or {}
    )


# Decorator for automatic error capture
def sentry_trace(operation_name: str):
    """Decorator to automatically trace operations and capture errors."""
    def decorator(func):
        def wrapper(*args, **kwargs):
            with sentry_sdk.start_transaction(op="function", name=operation_name):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    capture_exception_with_context(
                        e,
                        extra_context={
                            "function": func.__name__,
                            "operation": operation_name,
                            "args": str(args)[:200],  # Limit size
                            "kwargs": str(kwargs)[:200]
                        }
                    )
                    raise
        return wrapper
    return decorator

File: src/lib/sentry.ts
/**
 * Sentry configuration for CareBow frontend error monitoring
 */
import * as Sentry from '@sentry/react';
import { BrowserTracing } from '@sentry/tracing';
import { env } from './env';

// Environment configuration
const SENTRY_DSN = env.NEXT_PUBLIC_SENTRY_DSN;
const ENVIRONMENT = env.NEXT_PUBLIC_ENVIRONMENT || 'development';
const VERSION = env.NEXT_PUBLIC_APP_VERSION || '1.0.0';

/**
 * Initialize Sentry for error monitoring and performance tracking
 */
export function initSentry(): void {
  // Temporarily disable Sentry to debug the error
  console.info('Sentry initialization disabled for debugging');
  return;
  
  if (!SENTRY_DSN) {
    console.info('Sentry DSN not configured, skipping Sentry initialization');
    return;
  }

  Sentry.init({
    dsn: SENTRY_DSN,
    environment: ENVIRONMENT,
    release: VERSION,
    
    // Integrations
    integrations: [
      new BrowserTracing({
        // Simplified browser tracing without React Router instrumentation
        // to avoid import issues during initialization
      }),
    ],
    
    // Performance monitoring
    tracesSampleRate: ENVIRONMENT === 'development' ? 1.0 : 0.1,
    
    // Error sampling
    sampleRate: 1.0,
    
    // Additional configuration
    attachStacktrace: true,
    
    // Don't send personally identifiable information
    beforeSend(event, hint) {
      return filterSensitiveData(event, hint);
    },
    
    // Filter performance transactions
    beforeSendTransaction(event) {
      return filterPerformanceData(event);
    },
    
    // Configure allowed URLs (optional)
    allowUrls: [
      // Add your domain here in production
      /localhost/,
      /carebow\.com/,
    ],
    
    // Ignore certain errors
    ignoreErrors: [
      // Browser extensions
      'Non-Error promise rejection captured',
      'ResizeObserver loop limit exceeded',
      'Script error.',
      
      // Network errors
      'NetworkError',
      'Failed to fetch',
      
      // React development warnings
      'Warning: ',
    ],
  });

  // Set global tags
  Sentry.setTag('component', 'frontend');
  Sentry.setTag('framework', 'react');
  
  console.info(`Sentry initialized for environment: ${ENVIRONMENT}`);
}

/**
 * Filter sensitive data before sending to Sentry
 */
function filterSensitiveData(event: Sentry.Event, hint: Sentry.EventHint): Sentry.Event | null {
  // Remove sensitive information from the event
  if (event.request?.data) {
    const data = event.request.data;
    
    // Remove password fields
    if (typeof data === 'object' && data !== null) {
      const sanitized = { ...data };
      delete sanitized.password;
      delete sanitized.confirmPassword;
      delete sanitized.currentPassword;
      delete sanitized.newPassword;
      event.request.data = sanitized;
    }
  }
  
  // Remove sensitive headers
  if (event.request?.headers) {
    const headers = { ...event.request.headers };
    delete headers.Authorization;
    delete headers.Cookie;
    event.request.headers = headers;
  }
  
  // Add custom context
  event.tags = {
    ...event.tags,
    userAgent: navigator.userAgent,
    url: window.location.href,
  };
  
  return event;
}

/**
 * Filter performance data before sending to Sentry
 */
function filterPerformanceData(event: Sentry.Event): Sentry.Event | null {
  // Don't track certain routes in production
  if (ENVIRONMENT === 'production') {
    const transactionName = event.transaction;
    if (transactionName?.includes('/health') || transactionName?.includes('/ping')) {
      return null;
    }
  }
  
  return event;
}

/**
 * Capture exception with additional context
 */
export function captureException(
  error: Error,
  context?: {
    user?: { id: string; email?: string };
    extra?: Record<string, any>;
    tags?: Record<string, string>;
  }
): void {
  Sentry.withScope((scope) => {
    if (context?.user) {
      scope.setUser(context.user);
    }
    
    if (context?.extra) {
      Object.entries(context.extra).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
    }
    
    if (context?.tags) {
      Object.entries(context.tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }
    
    Sentry.captureException(error);
  });
}

/**
 * Capture message with additional context
 */
export function captureMessage(
  message: string,
  level: Sentry.SeverityLevel = 'info',
  context?: {
    user?: { id: string; email?: string };
    extra?: Record<string, any>;
    tags?: Record<string, string>;
  }
): void {
  Sentry.withScope((scope) => {
    if (context?.user) {
      scope.setUser(context.user);
    }
    
    if (context?.extra) {
      Object.entries(context.extra).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
    }
    
    if (context?.tags) {
      Object.entries(context.tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }
    
    Sentry.captureMessage(message, level);
  });
}

/**
 * Set user context for current session
 */
export function setUserContext(user: { id: string; email?: string; name?: string }): void {
  Sentry.setUser({
    id: user.id,
    email: ENVIRONMENT === 'development' ? user.email : undefined,
    username: user.name,
  });
}

/**
 * Add breadcrumb for debugging context
 */
export function addBreadcrumb(
  message: string,
  category: string = 'custom',
  level: Sentry.SeverityLevel = 'info',
  data?: Record<string, any>
): void {
  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data,
  });
}

/**
 * Clear user context (on logout)
 */
export function clearUserContext(): void {
  Sentry.setUser(null);
}
